master/internal/token/postgres_token.go

```diff
@@ -0,0 +1,154 @@
+// It takes a pointer to model.UserSession and modifies it.
+// It’s used to apply optional settings to the AccessToken object.
+type AccessTokenOption func(f *model.UserSession)
+
+// WithTokenExpiry function will add specified expiresAt (if any) to the access token table.
+func WithTokenExpiry(expiry *time.Duration) AccessTokenOption {
+	return func(s *model.UserSession) {
+		s.Expiry = CurrentTimeNowInUTC.Add(*expiry)
+	}
+}
+
+// WithTokenDescription function will add specified description (if any) to the access token table.
+func WithTokenDescription(description string) AccessTokenOption {
+	return func(s *model.UserSession) {
+		if description == "" {
+			return
+		}
+		s.Description = null.StringFrom(description)
+	}
+}
+
+// CreateAccessToken creates a new access token and store in
+// user_sessions db.
+func CreateAccessToken(
+	ctx context.Context, userID model.UserID, opts ...AccessTokenOption,
+) (string, model.TokenID, error) {
+	CurrentTimeNowInUTC = time.Now().UTC()
+	// Populate the default values in the model.
+	accessToken := &model.UserSession{
+		UserID:      userID,
+		CreatedAt:   CurrentTimeNowInUTC,
+		Expiry:      CurrentTimeNowInUTC.Add(DefaultTokenLifespan),
+		TokenType:   model.TokenTypeAccessToken,
+		Description: null.StringFromPtr(nil),
+		Revoked:     false,
+	}
+
+	// Update the optional ExpiresAt field (if passed)
+	for _, opt := range opts {
+		opt(accessToken)
+	}
+
+	var token string
+
+	err := db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		// AccessTokens should have a many:1 relationship with users.
+		// A new row is inserted into the user_sessions table, and the ID of the
+		// inserted row is returned and stored in user_sessions.ID.
+		_, err := tx.NewInsert().
+			Model(accessToken).
+			Column("user_id", "expiry", "created_at", "token_type", "revoked", "description").
+			Returning("id").
+			Exec(ctx, &accessToken.ID)
+		if err != nil {
+			return err
+		}
+
+		// A Paseto token is generated using the accessToken object and the private key.
+		v2 := paseto.NewV2()
+		privateKey := db.GetTokenKeys().PrivateKey
+		token, err = v2.Sign(privateKey, accessToken, nil)
+		if err != nil {
+			return fmt.Errorf("failed to generate user authentication token: %s", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return "", 0, err
+	}
+
+	return token, model.TokenID(accessToken.ID), nil
+}
+
+// AccessTokenUpdateOptions is the set of mutable fields for an Access Token record.
+type AccessTokenUpdateOptions struct {
+	Description *string
+	SetRevoked  bool
+}
+
+// UpdateAccessToken updates the description and revocation status of the access token.
+func UpdateAccessToken(
+	ctx context.Context, tokenID model.TokenID, options AccessTokenUpdateOptions,
+) (*model.UserSession, error) {
+	var tokenInfo model.UserSession
+	// Lock the token row for update.
+	err := db.Bun().NewSelect().For("UPDATE").Table("user_sessions").
```

Comment:
maybe i'm missing something, but this could be a single update query?
something like (pseudocode):
```
update user_sessions 
set revoked = options.Revoked
set description = options.Description
where
id = tokenID
and expiry < time.now()
and revoked is False
```

the only thing we'd miss is being able to print the error message `fmt.Errorf("unable to update inactive token with ID %v", tokenID)` since we'd just not find anything to update. is this the reason? if so, makes sense, just wondering!

Reply:
Yes, because we want the user to know that they cannot expect to update the token info (eg. description/lifespan) if the token is already expired. Hence we thought of keeping the Update and Select separately.

Reply:
> does select for update need to be in a transaction? what happens to the locked row if we don't end up updating?

Brought back transaction code, removed it by mistake.

Current behavior: When we run For("UPDATE"), it locks the rows being selected. This prevents other transactions from modifying those rows until your current transaction has finished. However, other transactions can still read the locked rows in many database systems unless explicitly restricted.